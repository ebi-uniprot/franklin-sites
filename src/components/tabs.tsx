import {
  useState,
  useCallback,
  useRef,
  Children,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import { v1 } from 'uuid';
import cn from 'classnames';
import { Except } from 'type-fest';

import '../styles/components/tabs.scss';

type TabProps = {
  /**
   * Title of that tab
   */
  title: ReactNode;
  /**
   * Optional ID for that tab, one of the expected options for the parent <Tabs> component
   */
  id?: string;
  /**
   * Content of that tab
   */
  children?: ReactNode;
  /**
   * Choose that tab as the default to be displayed
   */
  defaultSelected?: boolean;
  /**
   * Option to render and hide tab (display:none) rather than remove from the DOM
   */
  cache?: boolean;
} & Except<HTMLAttributes<HTMLDivElement>, 'title' | 'id'>;

// This is just a configuration component, it doesn't need to render anything as
// it will be used by a <Tabs> component
export const Tab = (_: TabProps) => null;

type TabsProps = {
  /**
   * <Tab> elements defining the content and title of each tab
   */
  children:
    | Array<ReactElement<TabProps> | null>
    | ReactElement<TabProps>
    | null;
  /**
   * Optional way of controling the tabs from the outside of this component by
   * assigning here a value corresponding to an 'id' prop of one of the child
   * <Tab>
   */
  active?: string | number;
} & Except<HTMLAttributes<HTMLDivElement>, 'children'>;

export const Tabs = ({ children, active, className, ...props }: TabsProps) => {
  const idRef = useRef(v1());

  const isManaged = typeof active !== 'undefined';

  // create an array of tab description objects out of the children's props
  const childrenArray = Children.toArray(children).filter(Boolean) as Array<
    ReactElement<TabProps>
  >;
  const tabs = childrenArray.map(
    // eslint-disable-next-line no-shadow
    ({ props: { id, ...props } }, index) => ({
      // set a default value for id depending on their index if needed
      id: id === undefined ? `${index}` : id,
      // and get the rest of the props as they are
      ...props,
    })
  );

  // state to use to decide which tab to render if this component is not managed
  const [selectedState, setSelectedState] = useState(() => {
    if (isManaged) {
      return active;
    }
    const defaultSelected = tabs.filter((tab) => tab.defaultSelected);
    if (defaultSelected.length) {
      if (defaultSelected.length > 1) {
        // eslint-disable-next-line no-console
        console.warn(
          `a <Tabs> component has been rendered with ${defaultSelected.length} <Tab defaultSelected> children. There should be a maximum of 1 default selected child.`
        );
      }
      return defaultSelected[0].id;
    }
    return tabs[0].id;
  });

  const handleClick = useCallback(
    (event: MouseEvent | KeyboardEvent) => {
      if (isManaged || !(event.currentTarget instanceof HTMLElement)) {
        return;
      }

      const { target } = event.currentTarget.dataset;
      // mouse click event, or if keyboard event, restrict to 'Enter' and spacebar keys
      if (
        event &&
        target !== undefined &&
        (!('key' in event) || event.key === 'Enter' || event.key === ' ')
      ) {
        setSelectedState(target);
      }
    },
    [isManaged]
  );

  const activeFromPropsOrState = isManaged ? active : selectedState;

  const selectedTab = tabs.find((tab) => tab.id === activeFromPropsOrState);
  if (!selectedTab) {
    throw new Error(`Could not find a tab with the id: "${selectedState}"`);
  }

  let content;
  const hasCacheTab = tabs.some(({ cache }) => cache);
  if (hasCacheTab) {
    content = tabs.map((tab) => {
      const selected = tab.id === selectedTab.id;
      return (
        (tab.cache || selected) && (
          <div key={tab.id} style={{ display: selected ? 'block' : 'none' }}>
            {tab.children}
          </div>
        )
      );
    });
  } else {
    content = selectedTab.children;
  }

  let unmanagedProps = {};
  // add event listeners in case this is not an externally managed component
  if (!isManaged) {
    unmanagedProps = {
      onClick: handleClick,
      onKeyPress: handleClick,
      tabIndex: 0,
    };
  }

  return (
    <div className={cn('tabs', className)} {...props}>
      <div className="tabs__header" role="tablist">
        {tabs.map(
          ({
            title,
            id,
            className, // eslint-disable-line no-shadow
            children: _,
            defaultSelected: __,
            cache: ___,
            ...props // eslint-disable-line no-shadow
          }) =>
            title && (
              <div
                key={id}
                data-testid="tab-title"
                data-target={id}
                role="tab"
                aria-controls={idRef.current}
                className={cn(
                  'tabs__header__item',
                  {
                    'tabs__header__item--active': id === activeFromPropsOrState,
                  },
                  className
                )}
                {...unmanagedProps}
                {...props}
              >
                {title}
              </div>
            )
        )}
      </div>
      <div role="tabpanel" id={idRef.current} data-testid="tab-content">
        {content}
      </div>
    </div>
  );
};
