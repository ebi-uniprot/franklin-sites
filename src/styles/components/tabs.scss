@import '../settings';
@import '../colours';

.tabs {
  &__header {
    width: 100%;
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    overflow-x: auto;
    box-shadow: inset 0 -0.1rem 0 0 var(--fr--color-platinum);
    margin-bottom: $global-margin;
    color: var(--fr--color-sapphire-blue);
    font-weight: 600;

    &__item {
      flex: 0 0 auto;
      padding: $global-padding;
      cursor: pointer;
      -webkit-user-select: none;
      -moz-user-select: none;
      -ms-user-select: none;
      user-select: none;

      &--bordered {
        border-right: 0.1rem solid var(--fr--color-platinum);
        border-top: 0.1rem solid var(--fr--color-platinum);
        border-radius: 0.2rem 0.2rem 0 0;
        background-color: initial;

        &:first-child {
          border-left: 0.1rem solid var(--fr--color-platinum);
        }
      }

      &--active,
      &:hover,
      &:focus {
        background-color: color(
          from var(--fr--color-pastel-blue) srgb r g b / 0.19
        );
        box-shadow: inset 0 -0.2rem 0 0 var(--fr--color-sea-blue);
      }

      &[aria-disabled='true'] {
        opacity: 0.5;
        user-select: none;
        cursor: not-allowed;
        background-color: initial;
        box-shadow: initial;

        &:active {
          pointer-events: none;
        }
      }
    }
  }
}

@supports (selector(:has(a))) {
  .tabs__header__item:has(> a) {
    padding: 0rem;

    & > a {
      padding: 1rem;
      display: block;
      color: var(--fr--color-sapphire-blue);
    }
  }
}
