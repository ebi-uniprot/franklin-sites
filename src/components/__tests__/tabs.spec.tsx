import { screen, render, fireEvent } from '@testing-library/react';

import { Tabs, Tab } from '../tabs';

jest.mock('uuid', () => ({
  v1: jest.fn(() => 'abcd'),
}));

describe('Tabs', () => {
  test('should render', () => {
    const { asFragment } = render(
      <Tabs>
        <Tab title={<div>Title 1</div>}>blah</Tab>
        <Tab title="Title 2">blaher</Tab>
        <Tab title="Title 3">blahest</Tab>
      </Tabs>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render other default tab', () => {
    const { asFragment } = render(
      <Tabs>
        <Tab title={<div>Title 1</div>}>blah</Tab>
        <Tab title="Title 2">blaher</Tab>
        <Tab title="Title 3" defaultSelected>
          blahest
        </Tab>
      </Tabs>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render tabs and pass extra props', () => {
    const { asFragment } = render(
      <Tabs data-value="value">
        <Tab title={<div>Title 1</div>} aria-label="blah!">
          blah
        </Tab>
        <Tab title="Title 2">blaher</Tab>
        <Tab title="Title 3" className="merged-class" data-key="3">
          blahest
        </Tab>
      </Tabs>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should show corresponding content when 2nd tab title is clicked', () => {
    render(
      <Tabs>
        <Tab id="a" title={<div>Title 1</div>}>
          blah
        </Tab>
        <Tab id="b" title="Title 2">
          blaher
        </Tab>
        <Tab id="c" title="Title 3">
          blahest
        </Tab>
      </Tabs>
    );
    let content = screen.queryByTestId('tab-content');
    expect(content).not.toHaveTextContent('blaher');
    const title = screen.queryAllByTestId('tab-title');
    fireEvent.click(title[1]);
    content = screen.queryByTestId('tab-content');
    expect(content).toHaveTextContent('blaher');
  });

  test('should show 2nd tab if defined as active', () => {
    render(
      <Tabs active="b">
        <Tab id="a" title={<div>Title 1</div>}>
          blah-1
        </Tab>
        <Tab id="b" title="Title 2">
          blaher
        </Tab>
        <Tab id="c" title="Title 3" />
      </Tabs>
    );
    const content = screen.queryByTestId('tab-content');
    expect(content).not.toHaveTextContent('blah-1');
    expect(content).toHaveTextContent('blaher');
  });
});
