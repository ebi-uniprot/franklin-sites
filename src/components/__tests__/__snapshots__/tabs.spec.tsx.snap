// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render 1`] = `
<DocumentFragment>
  <div
    class="tabs"
  >
    <div
      class="tabs__header"
      role="tablist"
    >
      <div
        aria-controls="abcd"
        class="tabs__header__item tabs__header__item--active"
        data-target="0"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        <div>
          Title 1
        </div>
      </div>
      <div
        aria-controls="abcd"
        class="tabs__header__item"
        data-target="1"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        Title 2
      </div>
      <div
        aria-controls="abcd"
        class="tabs__header__item"
        data-target="2"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        Title 3
      </div>
    </div>
    <div
      data-testid="tab-content"
      id="abcd"
      role="tabpanel"
    >
      blah
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs should render other default tab 1`] = `
<DocumentFragment>
  <div
    class="tabs"
  >
    <div
      class="tabs__header"
      role="tablist"
    >
      <div
        aria-controls="abcd"
        class="tabs__header__item"
        data-target="0"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        <div>
          Title 1
        </div>
      </div>
      <div
        aria-controls="abcd"
        class="tabs__header__item"
        data-target="1"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        Title 2
      </div>
      <div
        aria-controls="abcd"
        class="tabs__header__item tabs__header__item--active"
        data-target="2"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        Title 3
      </div>
    </div>
    <div
      data-testid="tab-content"
      id="abcd"
      role="tabpanel"
    >
      blahest
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs should render tabs and pass extra props 1`] = `
<DocumentFragment>
  <div
    class="tabs"
    data-value="value"
  >
    <div
      class="tabs__header"
      role="tablist"
    >
      <div
        aria-controls="abcd"
        aria-label="blah!"
        class="tabs__header__item tabs__header__item--active"
        data-target="0"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        <div>
          Title 1
        </div>
      </div>
      <div
        aria-controls="abcd"
        class="tabs__header__item"
        data-target="1"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        Title 2
      </div>
      <div
        aria-controls="abcd"
        class="tabs__header__item merged-class"
        data-key="3"
        data-target="2"
        data-testid="tab-title"
        role="tab"
        tabindex="0"
      >
        Title 3
      </div>
    </div>
    <div
      data-testid="tab-content"
      id="abcd"
      role="tabpanel"
    >
      blah
    </div>
  </div>
</DocumentFragment>
`;
